io-sim/src/Control/Monad/IOSim/Types.hs

```diff
@@ -615,6 +615,16 @@ instance MonadInspectMVar (IOSim s) where
         MVarEmpty _ _ -> pure Nothing
         MVarFull x _  -> pure (Just x)
 
+instance MonadTraceMVar (IOSim s) where
+  traceMVarIO _ (MVar mvar) f = traceTVarIO mvar traceMVarAsTVar
```

Comment:
The drawback of it is that everything is traced using `TraceTVar` constructor.  Could you add `TraceMVar` to `StmA` and use it here instead?

Reply:
Ignore my comment; `TraceTVar` constructor is only used to write to the `TVar` callback.